lib/IDE/CompletionInstance.cpp

```diff
@@ -586,6 +598,11 @@ bool CompletionInstance::shouldCheckDependencies() const {
   return threshold <= now;
 }
 
+void CompletionInstance::invalidateCachedCompilerInstance() {
+  std::lock_guard<std::mutex> lock(mtx);
+  CachedCIInvalidated = true;
```

Comment:
It looks to me that you have shared access of `CachedCIInvalidated` but without always using the mutex.
I think you can just make it a `std::atomic<bool>`.

Reply:
My bad, I moved 
```
  // Clear the cached CI if it's invalidated.
   if (CachedCIInvalidated) {
     clearCachedCompilerInstance();
   }
```
inside the mutex. Do you still think it should be `std::atomic<bool>`?

Reply:
I don't feel safe `CachedCIInvalidated` is mutated during the `performOperation()` anyways. Using mutex to guard it is still needed.

Reply:
Is the modification in `CompletionInstance::cacheCompilerInstance()` protected?

I would suggest it's better to use `std::atomic<bool>` because:
* One less thing to worry about accessing out of a mutex. With later changes you may end up either not protecting access to this field or blocking the modification of the new request on the mutex longer than you'd expect.
* It doesn't matter when this is getting modified or not, what matters is when you choose to check for it. It is like a "hint", e.g. say you had a "cancel" boolean that can be set while operation is running and then you check it at certain points, what matters is just to make it's use thread-safe.

How about renaming it to `CachedCIShouldBeInvalidated`, then it is more clear that modifying this doesn't have an effect until we check it at certain points.

Reply:
Yes `cacheCompilerInstance()` is private method and only called inside `performOperation()` after `std::lock_guard<std::mutex> lock(mtx)`.

Anyways, I agreed. Updated to use `std::atomic<bool>`